core/kernel/dt_driver_test.c

```diff
@@ -0,0 +1,529 @@
+			return TEE_ERROR_GENERIC;
+		}
+
+		dt_test_state.probe_clocks = SUCCESS;
+	}
+
+	if (IS_ENABLED(CFG_DRIVERS_RSTCTRL)) {
+		struct rstctrl *rstctrl0 = NULL;
+		struct rstctrl *rstctrl1 = NULL;
+		struct rstctrl *rstctrl = NULL;
+
+		DT_TEST_MSG("Probe reset controllers");
+
+		res = rstctrl_dt_get_by_index(fdt, node, 0, &rstctrl0);
+		if (res)
+			goto err_probe;
+
+		DT_TEST_MSG("Check valid reset controller");
+
+		if (rstctrl_assert(rstctrl0)) {
+			EMSG("Can't assert rstctrl %s", rstctrl_name(rstctrl0));
+			return TEE_ERROR_GENERIC;
+		}
+
+		res = rstctrl_dt_get_by_name(fdt, node, "rst0", &rstctrl);
+		if (res)
+			return res;
+
+		if (rstctrl != rstctrl0) {
+			EMSG("Unexpected reset controller reference");
+			return TEE_ERROR_GENERIC;
+		}
+
+		res = rstctrl_dt_get_by_name(fdt, node, "rst1", &rstctrl1);
+		if (res)
+			goto err_probe;
+
+		if (!rstctrl1 || rstctrl1 == rstctrl0) {
+			EMSG("Unexpected reset controller reference");
+			return TEE_ERROR_GENERIC;
+		}
+
+		dt_test_state.probe_resets = SUCCESS;
+	}
+
+	if (dt_test_state.probe_deferral != IN_PROGRESS) {
+		dt_test_state.probe_deferral = FAILED;
+		return TEE_ERROR_GENERIC;
+	}
+
+	dt_test_state.probe_deferral = SUCCESS;
+
+	return TEE_SUCCESS;
+
+err_probe:
+	assert(res);
+
+	if (res == TEE_ERROR_DEFER_DRIVER_INIT &&
+	    dt_test_state.probe_deferral == DEFAULT) {
+		/* We expect at least a probe deferral */
+		dt_test_state.probe_deferral = IN_PROGRESS;
+	}
+
+	return res;
+}
+
+static const struct dt_device_match dt_test_consumer_match_table[] = {
+	{ .compatible = "linaro,dt-test-consumer", },
+	{ }
+};
+
+DEFINE_DT_DRIVER(dt_test_consumer_driver) = {
+	.name = "dt-test-consumer",
+	.match_table = dt_test_consumer_match_table,
+	.probe = dt_test_consumer_probe,
+};
+
+static TEE_Result dt_test_crypt_consumer_probe(const void *fdt __unused,
+					       int node __unused,
+					       const void *compat_data __unused)
+{
+	TEE_Result res = dt_driver_get_crypto();
+	uint8_t __maybe_unused byte = 0;
+
+	if (res == TEE_ERROR_DEFER_DRIVER_INIT &&
+	    dt_test_state.crypto_dependencies == DEFAULT) {
+		/* We expect to be deferred */
+		dt_test_state.crypto_dependencies = IN_PROGRESS;
+	}
+
+	if (res)
+		return res;
+
+	if (dt_test_state.crypto_dependencies == DEFAULT) {
+		EMSG("Test expects at least a driver probe deferral");
+		dt_test_state.crypto_dependencies = FAILED;
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (crypto_rng_read(&byte, sizeof(byte))) {
+		dt_test_state.crypto_dependencies = FAILED;
+		return TEE_ERROR_GENERIC;
+	}
+
+	dt_test_state.crypto_dependencies = SUCCESS;
+	return TEE_SUCCESS;
+}
+
+static const struct dt_device_match dt_test_crypt_consumer_match_table[] = {
+	{ .compatible = "linaro,dt-test-crypt-consumer", },
+	{ }
+};
+
+DEFINE_DT_DRIVER(dt_test_consumer_driver) = {
+	.name = "dt-test-crypt-consumer",
+	.match_table = dt_test_crypt_consumer_match_table,
+	.probe = dt_test_crypt_consumer_probe,
+};
+
+#ifdef CFG_DRIVERS_CLK
```

Comment:
Why is this ifdef needed?

Reply:
Some flexibility is good. What happens without this ifdef, does something break?

Reply:
OK, let's keep it then.